### PR TITLE
[Tizen.Applications.RPCPort] Add RPCPort ParcelHeader APIs

### PR DESCRIPTION
**Pull request by upple**
Signed-off-by: ChangGyu Choi <uppletaste@gmail.com>

### Description of Change ###
RPCPort API needs to add parcel's header.
The follow APIs are added.

* Add new RPCPort API set ParcelHeader
* Add a disconnect port API.

### API Changes ###

 - ACR: https://code.sec.samsung.net/jira/browse/TCSACR-457

Adds:
 Class:
 - Tizen.Applications.RPCPort.ParcelHeader

 Operation:
 - Tizen.Applications.RPCPort.Disconnect()
 - Tizen.Applications.RPCPort.Parcel.GetHeader()
 - Tizen.Applications.RPCPort.ParcelHeader.SetTag()
 - Tizen.Applications.RPCPort.ParcelHeader.GetTag()
 - Tizen.Applications.RPCPort.ParcelHeader.SetSeqNum()
 - Tizen.Applications.RPCPort.ParcelHeader.GetSeqNum()
 - Tizen.Applications.RPCPort.ParcelHeader.GetTimeStamp()

### CODE REVIEW

**Comment by myroot** [comment]
Could you check TizenFX Desigin guidelines? 
https://github.sec.samsung.net/dotnet/DesignGuidelines

File: src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs [hunk 1]
```diff
@@ -18,6 +18,135 @@
...
+        /// <summary>
+        /// Constructor with TimeStamp.
+        /// </summary>
+        /// <exception cref="InvalidIOException">Thrown when an internal IO error occurs.</exception>
+        /// <since_tizen> 9 </since_tizen>
+        internal TimeStamp(Interop.LibRPCPort.Parcel.TimeStamp time)
+        {
+            second = time.sec;
+            nanoSecond = time.nsec;
+        }
```
Comment (myroot) [inline src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs:36]: 커플링은 줄일수록 좋을고 이렇게하면 외부와의 커플링을 0으로 할수 있습니다.
```suggestion
        internal TimeStamp(long second, long nanoSecond)
        {
            this.second = second;
            this.nanoSecond = nanoSecond;
        }
```

File: src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs [hunk 2]
```diff
@@ -18,6 +18,135 @@
...
+
+        /// <summary>
+        /// The second of TimeStamp.
+        /// </summary>
+        /// <since_tizen> 9 </since_tizen>
+        public long second
```
Comment (myroot) [inline src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs:42]: PascalCasing을 사용하는것이 룰입니다.

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions#capitalization-rules-for-identifiers

```suggestion
        public long Second
```

File: src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs [hunk 3]
```diff
@@ -18,6 +18,135 @@
...
+        {
+            get;
+            private set;
+        }
+
+        /// <summary>
+        /// The nano second of TimeStamp.
+        /// </summary>
+        /// <since_tizen> 9 </since_tizen>
+        public long nanoSecond
```
Comment (myroot) [inline src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs:52]:
```suggestion
        public long NanoSecond
```

File: src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs [hunk 4]
```diff
@@ -18,6 +18,135 @@
...
+            return tag;
+        }
+
+        /// <summary>
+        /// Sets sequence number of Header.
+        /// </summary>
+        /// <param name="seqNum">The seqence number of Header</param>
+        /// <exception cref="InvalidIOException">Thrown when an internal IO error occurs.</exception>
+        /// <since_tizen> 9 </since_tizen>
+        public void SetSeqNum(int seqNum)
```
Comment (myroot) [inline src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs:110]: ❌ DO NOT use abbreviations or contractions as part of identifier names.

https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/general-naming-conventions#using-abbreviations-and-acronyms

```suggestion
        public void SetSequenceNumber(int sequenceNumber)
```

File: src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs [hunk 5]
```diff
@@ -302,6 +431,19 @@ public byte[] Read(int size)
...
+        /// </summary>
+        /// <returns>Parcel header</returns>
+        /// <since_tizen> 9 </since_tizen>
+        public ParcelHeader GetHeader()
+        {
+            Interop.LibRPCPort.Parcel.GetHeader(_handle, out IntPtr handle);
+            ParcelHeader header = new ParcelHeader();
+            header._handle = handle;
+            return header;
+        }
```
Comment (myroot) [inline src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs:445]: 가능하면 Property를 사용해주세요, 
cache를 적용해서 매번 객체를 안만들게 해주세요

```suggestion
        ParcelHeader _header;
        public ParcelHeader Header
        {
            get 
            {
                 if (_header != null)
                     return _header;
                     
                 Interop.LibRPCPort.Parcel.GetHeader(_handle, out IntPtr handle);
                 _header = new ParcelHeader() {
                       _handle = handle
                 };
                 return _header;
            }
        }
```

**Review by myroot — CHANGES_REQUESTED** [review]

File: src/Tizen.Applications.Common/Interop/Interop.RPCPort.cs [hunk 6]
```diff
@@ -42,6 +42,26 @@ internal enum PortType
 
         internal static partial class Parcel
         {
+            /// <summary>
+            /// This structure represents the struct timespec.
+            /// </summary>
+            /// <since_tizen> 9 </since_tizen>
+            [StructLayout(LayoutKind.Sequential)]
+            internal struct TimeStamp
```
Comment (myroot) [inline src/Tizen.Applications.Common/Interop/Interop.RPCPort.cs:50]: 저희가 native 구조체로부터 marshing하기 위해 매핑하여 작성한 struct에 대해 빌드시 사이즈가 동일한지 검사를 하고 있습니다. 이를 위해 NativeStruct attribute를 작성해야합니다.
```suggestion
            [NativeStruct("struct timespec", Include = "time.h")]
            [StructLayout(LayoutKind.Sequential)]
            public struct TimeStamp
```

**Review by upple — COMMENTED** [review]
@myroot 말씀주신 내용 수정하였습니다.
감사합니다.

File: src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs [hunk 7]
```diff
@@ -18,13 +18,143 @@
...
+            if (r != Interop.LibRPCPort.ErrorCode.None)
+                throw new InvalidIOException();
+
+            return tag;
+        }
+
+        /// <summary>
+        /// Sets sequence number of Header.
+        /// </summary>
+        /// <param name="seqNum">The seqence number of Header</param>
```
Comment (myroot) [inline src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs:107]:
```suggestion
        /// <param name="sequenceNumber">The seqence number of Header</param>
```

**Review by myroot — APPROVED** [review]

File: src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs [hunk 8]
```diff
@@ -18,13 +18,143 @@
...
+    /// <summary>
+    /// This structure represents the time stamp.
+    /// </summary>
+    /// <since_tizen> 9 </since_tizen>
+    public class TimeStamp
+    {
+        /// <summary>
+        /// Constructor with TimeStamp.
+        /// </summary>
+        /// <exception cref="InvalidIOException">Thrown when an internal IO error occurs.</exception>
```
Comment (hjhun) [inline src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs:30]: I think this description is not needed.

Comment (upple) [reply src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs]: Done. Thanks!

File: src/Tizen.Applications.Common/Interop/Interop.Libc.cs [hunk 9]
```diff
@@ -17,11 +17,21 @@
...
     {
         [DllImport(Libraries.Libc, EntryPoint = "free", CallingConvention = CallingConvention.Cdecl)]
         internal static extern int Free(IntPtr ptr);
+
+        [NativeStruct("struct timespec", Include = "time.h")]
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct TimeStamp
+        {
+            public uint sec;
+            public uint nsec;
```
Comment (hjhun) [inline src/Tizen.Applications.Common/Interop/Interop.Libc.cs:34]: Is using uint okay?
uint size is unsigned-32bit integer.
The size of the nsec of the struct timespec will be 8bytes in the 64-bit machine.

File: src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Port.cs [hunk 10]
```diff
@@ -96,5 +96,17 @@ public void UnshareFile()
...
         }
+
+        /// <summary>
+        /// Disconnects the port.
+        /// </summary>
+        /// <exception cref="InvalidIOException">Thrown when an internal IO error occurrs.</exception>
+        /// <since_tizen> 9 </since_tizen>
+        public void Disconnect()
+        {
+            Interop.LibRPCPort.ErrorCode err = Interop.LibRPCPort.Port.Disconnect(Handle);
```
Comment (hjhun) [inline src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Port.cs:107]: The Interop.LibRPCPort.Port.Disconnect() returns the RPC_PORT_ERROR_INVALID_PARAMETER or the RPC_PORT_ERROR_NONE.
I think the exception should be changed to "ArgumentNullException".

Comment (myroot) [reply src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Port.cs:107]: ArgumentNullException should throw when argument is existed and it is wrong.
but this method there is no argment

Comment (hjhun) [reply src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Port.cs:107]: I got it. Thanks.

File: src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Port.cs [hunk 11]
```diff
@@ -96,5 +96,17 @@ public void UnshareFile()
             if (err != Interop.LibRPCPort.ErrorCode.None)
                 throw new InvalidIOException();
         }
+
+        /// <summary>
+        /// Disconnects the port.
+        /// </summary>
+        /// <exception cref="InvalidIOException">Thrown when an internal IO error occurrs.</exception>
```
Comment (myroot) [inline src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Port.cs:103]: There is no `InvalidIOException` in .net frameworks, did you intend `System.IO.IOException` or `System.InvalidOperationException`

File: src/Tizen.Applications.Common/Interop/Interop.Libc.cs (hunk 9 above)
Comment (myroot) [reply src/Tizen.Applications.Common/Interop/Interop.Libc.cs:34]: You should have two pair of struct and interop api for 32bit and 64bit
``` c#
[DllImport(Libraries.RpcPort, EntryPoint = "rpc_port_parcel_header_get_timestamp")]
 internal static extern ErrorCode GetTimeStamp32(IntPtr parcelHeaderHandle, ref Libc.TimeStamp32 time);

[DllImport(Libraries.RpcPort, EntryPoint = "rpc_port_parcel_header_get_timestamp")]
internal static extern ErrorCode GetTimeStamp64(IntPtr parcelHeaderHandle, ref Libc.TimeStamp64 time);
```

And you can check byte length on runtime like below
``` c#
bool is32bit = IntPtr.Size == 4;
bool is64bit = IntPtr.Size == 8;
```

Comment (myroot) [reply src/Tizen.Applications.Common/Interop/Interop.Libc.cs:34]: @WonyoungChoi Could you give a guide to prepare 64bit system? if not enough it.

File: src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Port.cs (hunk 10 above)
Comment (upple) [reply src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Port.cs:107]: I fixed it.

File: src/Tizen.Applications.Common/Interop/Interop.Libc.cs (hunk 9 above)
Comment (upple) [reply src/Tizen.Applications.Common/Interop/Interop.Libc.cs]: I changed to nint.
According to the reference, the type follows the system architecture.
https://docs.microsoft.com/ko-kr/dotnet/csharp/language-reference/builtin-types/nint-nuint
Please review them. Thanks

File: src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Port.cs (hunk 11 above)
Comment (upple) [reply src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Port.cs]: Done.

File: src/Tizen.Applications.Common/Interop/Interop.Libc.cs (hunk 9 above)
Comment (myroot) [reply src/Tizen.Applications.Common/Interop/Interop.Libc.cs]: awsome!

but I think you need to update .csproj to use language 9.0 spec like below
``` .csproj
  <PropertyGroup>
    <TargetFramework>netstandard2.0</TargetFramework>
    <LangVersion>9.0</LangVersion>
  </PropertyGroup>
```

Is there no error on build?

**Review by myroot — APPROVED** [review]

**Review by hjhun — APPROVED** [review]

File: src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs [hunk 12]
```diff
@@ -18,13 +18,141 @@
 
 namespace Tizen.Applications.RPCPort
 {
+    /// <summary>
+    /// This structure represents the time stamp.
```
Comment (wiertel) [inline src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs:22]: structure -> class
the time stamp -> the time stamp of an RPC Parcel

File: src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs [hunk 13]
```diff
@@ -18,13 +18,141 @@
...
+        /// </summary>
+        /// <since_tizen> 9 </since_tizen>
+        internal TimeStamp(long second, long nanoSecond)
+        {
+            this.Second = second;
+            this.NanoSecond = nanoSecond;
+        }
+
+        /// <summary>
+        /// The second of TimeStamp.
```
Comment (wiertel) [inline src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs:38]: second -> seconds, the name is ok, just change description

File: src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs [hunk 14]
```diff
@@ -18,13 +18,141 @@
...
+        /// </summary>
+        /// <since_tizen> 9 </since_tizen>
+        public long Second
+        {
+            get;
+            private set;
+        }
+
+        /// <summary>
+        /// The nano second of TimeStamp.
```
Comment (wiertel) [inline src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs:48]: second -> seconds

File: src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs [hunk 15]
```diff
@@ -18,13 +18,141 @@
...
+        /// <since_tizen> 9 </since_tizen>
+        public long NanoSecond
+        {
+            get;
+            private set;
+        }
+    }
+
+    /// <summary>
+    /// The class is the header that has the Parcel's information.
```
Comment (wiertel) [inline src/Tizen.Applications.Common/Tizen.Applications.RPCPort/Parcel.cs:59]: The class represents the header information of an RPC Parcel.

**Pull request merged**